### ISSUE TITLE
Speed Issue on Rails

### ISSUE THREAD

**Issue by francirp** [zurb/inky-rb#17]
I am experiencing a pretty serious speed issue when trying to render an inky template in Rails 3.2.

It seems to be related to the use of iteration within the template:

The following takes 8 seconds to render:

```erb
  <container>
    <% 20.times do |i| %>
      <wrapper>
        <row>
          <columns small="12">
            <p>Hello</p>
          </columns>
        </row>
        <row>
          <columns small="12">
            <h1>Hi</h1>
          </columns>
        </row>
      </wrapper>
      <wrapper>
        <row>
          <columns small="12">
            <p>Bonjour</p>
          </columns>
        </row>
      </wrapper>
    <% end %>
  </container>
```

Whereas this takes 67ms:

```html
  <container>
    <wrapper>
      <row>
        <columns small="12">
          <p>Hello</p>
        </columns>
      </row>
      <row>
        <columns small="12">
          <h1>Hi</h1>
        </columns>
      </row>
    </wrapper>
    <wrapper>
      <row>
        <columns small="12">
          <p>Bonjour</p>
        </columns>
      </row>
    </wrapper>
  </container>
```

**Comment by francirp** [comment]
@zokioki @kball any idea on this? Seems odd how long it takes...

**Comment by kball** [comment]
That's a great question... so the way inky and erb work together is first we parse the ERB, and then hand off to the inky parser... which right now uses REXML to parse the document as an XML doc and progressively modify it.

My guess is REXML is probably the source of the slowness... upon digging into it a bit, I've seen some articles talking about how crappy the performance is of REXML as compared go e.g. libxml or nokogiri.  I chose REXML since it is part of the standard library to minimize dependencies, but it may be too slow.  Looking at ruby toolbox https://www.ruby-toolbox.com/categories/html_parsing it looks like nokogiri is super widely used, so we probably can get away with introducing the dependency without making many people upset, what do you think?

**Comment by francirp** [comment]
@kball got it - that makes sense. Nokogiri is a pretty widely used XML parser - many well-known Rails gems depend on it - so I don't think people will have an issue with it being required as a dependency.

Ideally it would be nice if Inky first converted the inky template to HTML/ERB and then ran the ERB parser on the output, but I realize that may not be easy due to the ERB tags disrupting the XML tree. So I think you're right that the better first step would be to first speed up the XML parsing with Nokogiri.

Let me know how I can help and thanks for spearheading this project - it certainly solves a very big pain point for those of us who have been knee-deep in writing tables on tables on tables within email templates.

**Comment by kball** [comment]
Hi @francirp the main reason I'm doing the ERB first is because the set of potential transformations is much harder to reason about - ERB lets you use arbitrary logic, while Inky is a markup language with generally very well defined translation patterns.

Its not perfectly clean, there are a couple places where Inky is actually using significant logic and not just a simple markup transformation, bit it tends to be much more limited than you'd get with ERB.

In fact, there are cases where Inky generates different markup based on attributes, so I think doing it the other way wouldn't work very well, because you'd expect ERB to be able to set those attributes but doing so would break.

For moving to Nokogiri, I probably won't have time to look at it for a couple of weeks... if you'd like to take a crack at it and submit it as a PR, or even just start mucking around with it in a branch, I'd be happy to have you helping out.  It looks like the node manipulation methods within Nokogiri are actually pretty similar to REXML, so it may well be straightforward to just drop it in.  

There's a pretty comprehensive set of specs so you should be able to test out changes quickly as you go. I'd start by looking at the `release_the_kraken` method in lib/inky.rb and follow it down through the parsing/transformation path.

**Comment by francirp** [comment]
@kball I'll take a look and see what I can do. I think I understand how this is working. I'll keep you posted if I run into questions / problems.

**Comment by francirp** [comment]
https://github.com/zurb/inky-rb/pull/19

**Comment by kball** [comment]
Fixed by #19
